### PR TITLE
Enum feature

### PR DESCRIPTION
**Pull request by queeksm**
Modifications for the Enum Methods to co0mply with the project specifications.

### CODE REVIEW

File: my_enums.rb
Comment (Forison) [inline my_enums.rb:22]: * [ ] ``` my_each_with_index ``` Must return enumerator when no block is given. the code below return nothing
``` array = [1,2,5,4,7]
array.my_each_with_index```

File: my_enums.rb
Comment (Forison) [inline my_enums.rb:179]: * [x] The two statements below ought to return ```true``` and ```false``` as it does in ruby's ```none?``` method  
```
 %w{ant bear cat}.my_none?(/d/)                        #=> true
 [1, 3.14, 42].my_none?(Float)                         #=> false 
```

File: my_enums.rb
Comment (Forison) [inline my_enums.rb:92]: * [ ] This ought to return FALSE and TRUE respectively as in ruby's ```all?```
```
# puts %w[ant bear cat].my_all?(/t/)                        #=> false
# puts [1, 2i, 3.14].my_all?(Numeric)                       #=> true
```

File: my_enums.rb
Comment (Forison) [inline my_enums.rb:231]: * [ ] my_map is should be equivalent to ruby's ```map``` method
refers to [ruby doc ](https://ruby-doc.org/core-2.7.0/Enumerable.html#method-i-count)

**Review by Forison — CHANGES_REQUESTED** [review]
### status: require changes🛑
# Hi Andres,
You have done a good job so far.
#### ⬇️Below are what we have to do to get this project approved⬇️
## Kindly make these changes and submit the project for review🙂 
Happy coding😇
Addo Forison

File: my_enums.rb
Comment (stratospherique) [inline my_enums.rb]: * [ ] enumerables my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class
```
    array = [1,2,5,'4',7]
    p array.my_any?(Integer) == array.any?(Integer) # false
```

* [ ] enumerables my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex
  ```
    array = %w[dog door rod blade]
    array[1] = 9
    p array.my_any?(/d/) == array.any?(/d/) # false
  ```

* [ ] enumerables my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
  ```
    array = %w[dog door rod blade]
    p array.my_any?('dog') == array.any?('dog') # false
  ```

File: my_enums.rb
Comment (stratospherique) [inline my_enums.rb:278]: * [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block
  ```ruby
    range = Range.new(5, 50)
    p range.my_inject(4) { |prod, n| prod * n } # returns this error
    #NoMethodError:
    #    undefined method `[]' for 5..50:Range
  ```

* [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
  ```ruby
    range = Range.new(5, 50)
    p range.my_inject(2, :*) # returns this error
    #NoMethodError:
    #    undefined method `[]' for 5..50:Range
  ```

File: my_enums.rb
Comment (stratospherique) [inline my_enums.rb]: should have working `my_each_with_index` **(covered by automated tests)** identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
  - my_each_with_index calls the given block once for each element in self

The second argument of **yield** is supposed to be the index of the element in regards to the whole array. The method index that you are passing as a second argument is not giving the right index :pray: fix it

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great work implementing the suggested changes :clap: However, some methods requires further refactoring in order to get approval for the project. check the code comments :arrow_down: 

### That's it :relaxed: When finished, kindly resubmit another code review request :pray: 

Happy coding :v: 
_**Ahmed Mahfoudh**_

File: my_enums.rb
Comment (majovanilla) [inline my_enums.rb:17]: You didn't have to delete the code, but to add a new line to it. Your my_each method is not working correctly now. It should call the given block once for each element. Try putting a line of code with the element and the index and check the result `puts some_array.my_each_with_index { | ele, index | puts "#{ele}: #{index}"  }`

File: my_enums.rb
Comment (majovanilla) [inline my_enums.rb]: - [ ] should have working `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- Your method doesn't consider the possibility when a string is given as an argument `(some_enumerable.my_any?('cow')).`

File: my_enums.rb
Comment (majovanilla) [inline my_enums.rb]: - [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- Your method should work with blocks as arguments. Right now, it has a problem when you do something like this: `stored_array.my_inject{ |sum, n| sum + n } ` It returns a different value (I think it's adding the first element twice)

- Your method should work with symbols. Right now, it has a problem when you do something like this: `stored_array.my_inject(:+) ` It returns a different value (I think it's adding the first element twice).

It's very important that you use a stored array to try to reproduce the error and call the method multiple times on the same array, in case you are mutating the original array.

**Review by majovanilla — CHANGES_REQUESTED** [review]
Hey @queeksm , 👋

## Good job on this project! 👍

There are still some changes 👇 you need to implement to approve this request 🔍 Please check the commented code below.

## Please, make all REQUIRED changes and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack (@majo)
I would be happy to help! 👩‍💻 


Happy coding 🤓

--Maria Reyes

File: my_enums.rb
Comment (queeksm) [reply my_enums.rb:17]: Tested the enum with the code provided and it behaved exactly as the one in the enum library on ruby, i can't replicate the problem that you saw. I would be grateful if you could give some additional insight of what is wrong, maybe some screenshots.

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello coders,
Thanks for the improvements done!
However, there are a few bugs to fix before moving forward. Take a look at these result from testing your methods

1.
```ruby
2.6.5 :002 > words = %w[dog door rod blade]
 => ["dog", "door", "rod", "blade"] 
2.6.5 :003 > words.my_any?('cat')
Traceback (most recent call last):
       10: from /home/maelfosso/.rvm/rubies/ruby-2.6.5/bin/irb:23:in `<main>'
        9: from /home/maelfosso/.rvm/rubies/ruby-2.6.5/bin/irb:23:in `load'
        8: from /home/maelfosso/.rvm/rubies/ruby-2.6.5/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        7: from (irb):3
        6: from my_enums.rb:109:in `my_any?'
        5: from my_enums.rb:14:in `my_each'
        4: from my_enums.rb:14:in `times'
        3: from my_enums.rb:16:in `block in my_each'
        2: from my_enums.rb:110:in `block in my_any?'
        1: from my_enums.rb:110:in `=~'
TypeError (type mismatch: String given)
2.6.5 :004 > words.any?('cat')
 => false 
2.6.5 :005 > 
```
2.
```ruby
2.6.5 :039 > array
 => [99, 6, 57, 43, 64, 16, 44, 53, 11, 22, 92, 94, 23, 29, 11, 84, 8, 82, 8, 57, 10, 26, 67, 40, 17, 85, 22, 88, 19, 33, 68, 57, 52, 19, 25, 88, 93, 44, 11, 32, 42, 65, 43, 10, 72, 64, 85, 77, 8, 89, 28, 57, 56, 82, 65, 4, 4, 9, 1, 26, 73, 56, 93, 54, 38, 28, 61, 99, 1, 99, 59, 50, 57, 86, 13, 47, 25, 73, 64, 31, 35, 57, 2, 24, 96, 17, 7, 68, 49, 44, 27, 56, 91, 33, 86, 10, 67, 29, 4, 16] 
2.6.5 :040 > my_each_output = ''
 => "" 
2.6.5 :041 >       block = proc { |num, idx| my_each_output += "Num: #{num}, idx: #{idx}\n" }
 => #<Proc:0x000055db32a30300@(irb):41> 
2.6.5 :042 >       array.each_with_index(&block)
 => [99, 6, 57, 43, 64, 16, 44, 53, 11, 22, 92, 94, 23, 29, 11, 84, 8, 82, 8, 57, 10, 26, 67, 40, 17, 85, 22, 88, 19, 33, 68, 57, 52, 19, 25, 88, 93, 44, 11, 32, 42, 65, 43, 10, 72, 64, 85, 77, 8, 89, 28, 57, 56, 82, 65, 4, 4, 9, 1, 26, 73, 56, 93, 54, 38, 28, 61, 99, 1, 99, 59, 50, 57, 86, 13, 47, 25, 73, 64, 31, 35, 57, 2, 24, 96, 17, 7, 68, 49, 44, 27, 56, 91, 33, 86, 10, 67, 29, 4, 16] 
2.6.5 :043 >       each_output = my_each_output.dup
 => "Num: 99, idx: 0\nNum: 6, idx: 1\nNum: 57, idx: 2\nNum: 43, idx: 3\nNum: 64, idx: 4\nNum: 16, idx: 5\nNum: 44, idx: 6\nNum: 53, idx: 7\nNum: 11, idx: 8\nNum: 22, idx: 9\nNum: 92, idx: 10\nNum: 94, idx: 11\nNum: 23, idx: 12\nNum: 29, idx: 13\nNum: 11, idx: 14\nNum: 84, idx: 15\nNum: 8, idx: 16\nNum: 82, idx: 17\nNum: 8, idx: 18\nNum: 57, idx: 19\nNum: 10, idx: 20\nNum: 26, idx: 21\nNum: 67, idx: 22\nNum: 40, idx: 23\nNum: 17, idx: 24\nNum: 85, idx: 25\nNum: 22, idx: 26\nNum: 88, idx: 27\nNum: 19, idx: 28\nNum: 33, idx: 29\nNum: 68, idx: 30\nNum: 57, idx: 31\nNum: 52, idx: 32\nNum: 19, idx: 33\nNum: 25, idx: 34\nNum: 88, idx: 35\nNum: 93, idx: 36\nNum: 44, idx: 37\nNum: 11, idx: 38\nNum: 32, idx: 39\nNum: 42, idx: 40\nNum: 65, idx: 41\nNum: 43, idx: 42\nNum: 10, idx: 43\nNum: 72, idx: 44\nNum: 64, idx: 45\nNum: 85, idx: 46\nNum: 77, idx: 47\nNum: 8, idx: 48\nNum: 89, idx: 49\nNum: 28, idx: 50\nNum: 57, idx: 51\nNum: 56, idx: 52\nNum: 82, idx: 53\nNum: 65, idx: 54\nNum: 4, idx: 55\nNum: 4, idx: 56\nNum: 9, idx: 57\nNum: 1, idx: 58\nNum: 26, idx: 59\nNum: 73, idx: 60\nNum: 56, idx: 61\nNum: 93, idx: 62\nNum: 54, idx: 63\nNum: 38, idx: 64\nNum: 28, idx: 65\nNum: 61, idx: 66\nNum: 99, idx: 67\nNum: 1, idx: 68\nNum: 99, idx: 69\nNum: 59, idx: 70\nNum: 50, idx: 71\nNum: 57, idx: 72\nNum: 86, idx: 73\nNum: 13, idx: 74\nNum: 47, idx: 75\nNum: 25, idx: 76\nNum: 73, idx: 77\nNum: 64, idx: 78\nNum: 31, idx: 79\nNum: 35, idx: 80\nNum: 57, idx: 81\nNum: 2, idx: 82\nNum: 24, idx: 83\nNum: 96, idx: 84\nNum: 17, idx: 85\nNum: 7, idx: 86\nNum: 68, idx: 87\nNum: 49, idx: 88\nNum: 44, idx: 89\nNum: 27, idx: 90\nNum: 56, idx: 91\nNum: 91, idx: 92\nNum: 33, idx: 93\nNum: 86, idx: 94\nNum: 10, idx: 95\nNum: 67, idx: 96\nNum: 29, idx: 97\nNum: 4, idx: 98\nNum: 16, idx: 99\n" 
2.6.5 :044 >       my_each_output = ''
 => "" 
2.6.5 :045 >       array.my_each_with_index(&block)
 => [99, 6, 57, 43, 64, 16, 44, 53, 11, 22, 92, 94, 23, 29, 11, 84, 8, 82, 8, 57, 10, 26, 67, 40, 17, 85, 22, 88, 19, 33, 68, 57, 52, 19, 25, 88, 93, 44, 11, 32, 42, 65, 43, 10, 72, 64, 85, 77, 8, 89, 28, 57, 56, 82, 65, 4, 4, 9, 1, 26, 73, 56, 93, 54, 38, 28, 61, 99, 1, 99, 59, 50, 57, 86, 13, 47, 25, 73, 64, 31, 35, 57, 2, 24, 96, 17, 7, 68, 49, 44, 27, 56, 91, 33, 86, 10, 67, 29, 4, 16] 
2.6.5 :046 > each_output == my_each_output
 => false 
2.6.5 :047 > 
```

Kindly fix those bugs and submit a new code review request
Thanks

Mael FOSSO

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi @queeksm      
#### Great Job so far, Just few things to work on. 
### General
- [ ] `my_any?`:
- when no block or argument is given returns true if at least one of the collection is not false or nil
`p [nil, false, nil, false].my_any?` - should return `false`
- [ ] You should have only one `my_map` method, am seeing you implemented four.
### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great work implementing the suggested changes :clap: :clap: However, one method needs further refactoring in order to get approval for this project. Check :arrow_down: 

- **[REQUIRED]** should have working `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
* [ ] my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
  ```ruby
    array = [1,true,20,'dom']
    p array.my_any? == array.any? # false
  ```


### That's it :relieved: When finished, kindly submit a code re-review request :pray: 
**_Ahmed Mahfoudh_**

**Review by pelzolga123 — APPROVED** [review]
Approved.

**Comment by pelzolga123** [comment]
## Great work, congratulation :+1: Project Approved.
You did all changes required previously. Before you merge this Pull Request, please kindly change the next two things:
*  `multiply_els` method should be outside Enumerable module. 
* `The Project Brief` section in Readme file says that this task is a part of the "HTML and CSS course". Please change it to the "Ruby course".  

Then go ahead and merge.
_Goodluck in your next project_ *Happy coding*

**Pull request merged**
